### DISCUSSION TITLE
New lines and formatting in meta shortcodes

### DISCUSSION THREAD

**Discussion by royfrancis** [quarto-dev/quarto-cli#7228] [Q&A]
### Description

````
---
title: Metadata formatting

block-title: Title
block-a: "Here we have an image. \n![](https://quarto.org/quarto.png)"
block-b: "Here is a call-out. :::{.callout-note} \nThis is a call-out. \n:::"
---

## {{< meta block-title >}}

{{< meta block-a >}}

{{< meta block-b >}}

````

![Screenshot 2023-10-14 at 17 56 01](https://github.com/quarto-dev/quarto-cli/assets/5083500/be0d9c91-c437-467c-b31d-beb2cf046d75)

```
▸ quarto --version
1.3.433
```

**Comment by cscheid** [comment]
That's a bug, but it's one that it'll be hard for us to fix in general, because it ultimately involves a recursive call into the quarto pipeline, and tricky issues about when to resolve what kinds of syntax. I think we only resolve markdown inlines (and no custom nodes).

**Reply by mcanouil** [reply]
Is it though? Are `\n` supposed to be evaluated?
One should use YAML multiline syntax to write multiple lines in my opinion.

The only issue I see, is that the callout is not taken into account.

````qmd
---
title: Metadata formatting

block-title: Title
block-a: |
  Here we have an image.  

  ![](https://quarto.org/quarto.png)
block-b: |
  Here is a call-out.

  ::: {.callout-note}
  This is a call-out.
  :::
---

## {{< meta block-title >}}

{{< meta block-a >}}

{{< meta block-b >}}
````

<img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/a65d29d2-4f4b-4d85-8e2b-a0e68d9c3a5c">

**Reply by cscheid** [reply]
> because it ultimately involves a recursive call into the quarto pipeline, and tricky issues about when to resolve what kinds of syntax. 

Callouts being one of the kinds of syntax that "it's too late to resolve" by the time meta shortcodes are evaluated.

**Reply by cscheid** [reply]
> Are \n supposed to be evaluated?

There was a single `\n` in the original block, which is not enough to break paragraphs. I think that was the cause of it.

**Reply by royfrancis** [reply]
What about simple markdown tables? Are they supposed to work in meta shortcodes?

```
---
table: |
  |Row|Col|
  |:---|:---|
  |A|1|
  |B|2|
---

{{< meta table >}}
```

**Reply by mcanouil** [reply]
What's your goal exactly with all this?

**Reply by royfrancis** [reply]
I probably should have started with that :) I am writing tutorials to run the same analysis using three different toolkits. So I have three different qmd files with different code, but the text and explanations will be same. So, to keep the text in sync and avoid duplication, I thought of simply replacing them with meta variables. I have one big meta variable file with all the text blocks. And the text will of course consist of anything from headings to tables, call outs etc. That's where this question comes in. And it's not just 3 files, there are many sets of three.

Here are the old materials which was put together using Rmd and jupyter: https://nbisweden.github.io/workshop-scRNAseq/exercises. I have the task of upgrading this to quarto.

**Reply by mcanouil** [reply]
If you are using `knitr` as the main engine, I suggest to use parameters and the parameters file which is a YAML file.

Also, why not write partials document and use the include shortcode?

It seems you are making your life harder that it needs to be.

**Reply by royfrancis** [reply]
I started with parameters but params cannot be accessed from meta shortcodes (#7149). Or perhaps you meant to eval code chunks to generate the text which I think is more complex than shortcodes.

Two documents will use knitr and one uses jupyter engine.

By partials, do you mean _child qmds? I haven't worked with it, so I don't know anything about it. Wouldn't that require many different child qmds? So if there are 15 paragraphs, would there be 15 child documents?

**Reply by cderv** [reply]
> By partials, do you mean _child qmds? I haven't worked with it, so I don't know anything about it. Wouldn't that require many different child qmds? So if there are 15 paragraphs, would there be 15 child documents?

I believe it means small reusable document that Quarto can easily include : https://quarto.org/docs/authoring/includes.html

This would be IMO also simpler that your `meta` usage. Yes it would mean 15 files if you have 15 different inclusion, but it would be inserted correctly at the right place in the rendering, and seems to better to use for what you want that meta shortcode. Especially for complew content like `table` example you show.

**Reply by royfrancis** [reply]
I have always thought of child documents for something more substantial than a small paragraph. Having 15 or 20 separate documents sounds like a hassle to manage and update, especially when most of them are just simple text. I feel like child docs are overkill for my use case. Besides, meta shortcodes work well so far, except complex formatting.

I suppose I could have a mix: meta variables for simple text and child qmd for the few complex blocks. Or just leave the complex formatted part directly in the parent qmd.

**Reply by cderv** [reply]
> Having 15 or 20 separate documents sounds like a hassle to manage and update, especially when most of them are just simple text. I feel like child docs are overkill for my use case.

Currently you are managing those 15 / 20 content in meta fields right ? Is it easier because they are all in the same place ? 
Because it is still 15 / 20 text to manage. 

Personally for long markdown block I need to insert, I find files better, even lots of small ones, because I gain formatting (!!) and this is quite important to me in edition. So with good IDE, and some organisation I find it not so much hassle to have 
* One folder for shared content, e.g in a `_shared/` folder (not the `_` so that they are not rendered)
* Then have all the shared content in some files to easily include with the quarto shortcodes 

Usually I find maintaining some markdown formatted text in YAML header field really hard, compared to ease of formatting in IDE (even using a visual editor tool in VSCODE or rstudio ide)

Anyhow, YAML multiline would be the way otherwise, but some feature won't work as mentioned (like callouts)
